proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,429 @@
+    
+Effect and Interactions
+-----------------------
+
+* Modules are now compositional.
+
+* The examples in the Motivation_ section are accepted.
+
+* There is a potential ambiguity between local modules and top-level modules. In particular, this might
+  happen between the implicit local module of a type declaration and a top-level module. For example::
+
+    -- top of file:
+    module A where
+
+    import qualified T ( x )
+
+    data T = MkT { x :: Int }
+
+    y = T.x
+
+  There will be two identifiers ``T.x`` in scope: both the one imported from ``T`` and the record selector
+  in the type ``T``. This situation will lead to an error, as do other sources of ambiguity.
+
+* The ability to detect dependencies of a module by parsing only a prefix of the module is retained.
+  Local modules are always imported only by ``import module``, never plain ``import``. Plain ``import``
+  statements remain at the top of the file.
+
+* Other than corner cases around ambiguity, this proposal is backward compatible; it is not "fork-like".
+
+* Proposal `#160`_ allows users to suppress field selectors, thus ameliorating a small part
+  of what has motivated this proposal.
+
+Costs and Drawbacks
+-------------------
+
+* This is a significant new bit of implementation and specification, and it should require the
+  requisite level of support from the community to be accepted.
+
+Alternatives
+------------
+
+There are many alternatives. The list below is shamelessly cribbed from `#205`_.
+
+* Proposal `#205`_. That proposal essentially tweaks the ``qualified`` feature to
+  become more flexible and exportable. It has proved hard to digest (from the commentary),
+  though, and solves fewer problems than this proposal. On the other hand, it is likely
+  easier to implement.
+
+  This proposal is inspired by some of the topics that came up in the conversation
+  for `#205`_, and I'm grateful for @deepfire's efforts on that proposal.
+
+* 2005 Coutts, ``as`` in export lists: `<https://mail.haskell.org/pipermail/libraries/2005-March/003390.html>`_ . Salient points:
+  letting modules export other modules' contents qualified with the module name`
+  
+* 2006 Wallace, explicit namespaces for module names: `<https://ghc.haskell.org/trac/ghc/wiki/Commentary/Packages/PackageNamespacesProposal>`_ . Salient points:
+  The declaration import namespace brings into availability the subset of the hierarchy of module names rooted in the package "foo-1.3", at the position ``Data.Foo``
+  
+* 2013 de Castro Lopo, qualified exports: `<https://wiki.haskell.org/GHC/QualifiedModuleExport>`_
+  ``qualified module T`` in export list and is essentially a subset of this proposal.
+
+Future Work
+-----------
+
+I see a few future directions along these lines, but I leave it to others to flesh these out.
+
+1. We can imagine *parameterized local modules*, where all the functions defined therein share
```

Comment:
In particular, I think this proposal is a common subset of almost any serious parameterized module proposal, so is a pretty reasonable thing to implement.